### ISSUE TITLE
Sum of shap values does not equal to the prediction (again)

### ISSUE THREAD

**Issue by pseudotensor** [catboost/catboost#2136]
Repro was provided: https://github.com/catboost/catboost/issues/1125#issuecomment-1023077820

https://github.com/catboost/catboost/issues/1125 original issue  not fixed

**Comment by ek-ak** [comment]
Hello!  Sorry for the long wait. We are investigating problem now.
Hotfix: converting model to 'json' format and loading it back from 'json' solves problem.
```
model.save_model('shap_model.json', format='json')
model = catboost.CatBoost().load_model('shap_model.json', format='json')
```

**Comment by fluid-gun** [comment]
> Hello! Sorry for the long wait. We are investigating problem now. Hotfix: converting model to 'json' format and loading it back from 'json' solves problem.
> 
> ```
> model.save_model('shap_model.json', format='json')
> model = catboost.CatBoost().load_model('shap_model.json', format='json')
> ```

Hi @ek-ak, your hotfix did not solve the problem. Any updates?

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
hi @fluid-gun ,  could please provide a reproducer?
that from https://github.com/catboost/catboost/issues/1125 works correctly for https://github.com/catboost/catboost/files/7949738/catshapproblem2.zip

```python
import numpy as np
import pickle
from catboost import CatBoost, CatBoostClassifier, CatBoostRegressor, EFstrType, Pool

model = CatBoostRegressor()
X, y, cat_features = pickle.load(open("catshapproblem.pkl", "rb"))

model.load_model("catshapproblem.model")
preds_src = model.predict(
        data=X,
        prediction_type="RawFormulaVal",
        ntree_start=0,
    )

model.save_model('shap_model.json', format='json')   # trick catboost to set correct scale and bias
model = CatBoost().load_model('shap_model.json', format='json')
preds_json = model.predict(
        data=X,
        prediction_type="RawFormulaVal",
        ntree_start=0,
    )
assert np.isclose(preds_src, preds_json).all(), "src and json predictions differ"

data = Pool(X, label=y, cat_features=cat_features)
preds_shap = model.get_feature_importance(
    data=data,
    type=EFstrType.ShapValues
)
preds_raw = model.predict(
        data=X,
        prediction_type="RawFormulaVal",
        ntree_start=0,
    )

max_abs_error = np.max(np.fabs(preds_raw - np.sum(preds_shap, axis=1)))
assert np.isclose(preds_raw, np.sum(preds_shap, axis=1)).all(), "catboost shapley does not sum up correctly: %g" % max_abs_error
```

**Comment by fluid-gun** [comment]
@Evgueni-Petrov-aka-espetrov my bad, found there was a wrong step in the preprocessing. btw, is there a more permanent fix to the bias introduction problem? or is the hotfix all we have at the moment

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
the issue is not fixed yet ☹️ 
though it seems that @ek-ak has found the root-cause
we have an internal issue for this problem -- i will raise the priority

**Comment by Samantha-movius** [comment]
Hi there, I'm facing the same issue with the SHAP values not adding up to the prediction value. @ek-ak when I apply the hotfix solution above, it changes the model's predicted value to match the SHAP values (as opposed to changing the SHAP values to match the original prediction which I expected). 

This change dramatically changes the scale of the predictions (insurance claim amounts) that were initially in the range (0, 67000) to (-4, 11). Is this expected?

**Comment by tilak-trinity** [comment]
Hello- Wanted to checkif there is any development in this issue.
I am trying to check the additivity using worked out examples in shap page.

`import xgboost

import shap

# train an XGBoost model
X, y = shap.datasets.california()
model = xgboost.XGBRegressor().fit(X, y)

# explain the model's predictions using SHAP
# (same syntax works for LightGBM, CatBoost, scikit-learn, transformers, Spark, etc.)
explainer = shap.Explainer(model)
shap_values = explainer(X)

shap_values_df = pd.DataFrame(shap_values.values, columns = X.columns.to_list())
shap_values_df['shap_sum']= shap_values_df.sum(axis=1, numeric_only=True)
shap_values_df['shap_additivity']=shap_values_df['shap_sum']+shap_values.base_values
shap_values_df['y']=y`

**Comment by frehbach** [comment]
Possibly this helps someone else with the issue:
For us the issue was actually using virtual_ensembles_predict of catboost instead of the usual predict method. 
As the prediction is then the mean of multiple predictions of the chosen ensembles it does not equal the "standard" prediction with which the shapley values are calculated. 
Switching to the usual catboost.predict resulted in the shapley values correctly adding up again.
